src/merkle/new_smt/mod.rs

```diff
@@ -0,0 +1,272 @@
+            let node = EmptySubtreeRoots::entry(self.depth(), index.depth() + 1);
+
+            InnerNode { left: *node, right: *node }
+        })
+    }
+
+    fn insert_inner_node(&mut self, index: NodeIndex, inner_node: InnerNode) {
+        self.inner_nodes.insert(index, inner_node);
+    }
+
+    fn insert_value(&mut self, key: Self::Key, value: Self::Value) -> Option<Self::Value> {
+        let leaf_index: LeafIndex<NEW_SMT_DEPTH> = key.into();
+        match self.leaves.get_mut(&leaf_index.value()) {
+            Some(leaf) => match leaf {
+                NewSmtLeaf::Single(kv_pair) => {
+                    // if the key is already in this entry, update the value and return
+                    if kv_pair.0 == key {
+                        let old_value = kv_pair.1;
+                        kv_pair.1 = value;
+                        return Some(old_value);
+                    }
+
+                    // transform the entry into a list entry, and make sure the key-value pairs
+                    // are sorted by key
+                    let mut pairs = vec![*kv_pair, (key, value)];
+                    pairs.sort_by(|(key_1, _), (key_2, _)| cmp_keys(*key_1, *key_2));
+
+                    self.leaves.insert(leaf_index.value(), NewSmtLeaf::Multiple(pairs));
+
+                    None
+                }
+                NewSmtLeaf::Multiple(kv_pairs) => {
+                    match kv_pairs.binary_search_by(|kv_pair| cmp_keys(kv_pair.0, key)) {
+                        Ok(pos) => {
+                            let old_value = kv_pairs[pos].1;
+                            kv_pairs[pos].1 = value;
+
+                            Some(old_value)
+                        }
+                        Err(pos) => {
+                            kv_pairs.insert(pos, (key, value));
+
+                            None
+                        }
+                    }
+                }
+            },
+            None => {
+                self.leaves.insert(leaf_index.value(), NewSmtLeaf::Single((key, value)));
+
+                Some(Self::Value::default())
+            }
+        }
+    }
+
+    fn get_leaf(&self, key: &Self::Key) -> Self::Leaf {
+        let leaf_pos = LeafIndex::<NEW_SMT_DEPTH>::from(*key).value();
+
+        match self.leaves.get(&leaf_pos) {
+            Some(leaf) => leaf.clone(),
+            None => NewSmtLeaf::Single((
+                *key,
+                Word::from(*EmptySubtreeRoots::entry(self.depth(), self.depth())),
+            )),
+        }
+    }
+
+    fn hash_leaf(leaf: &Self::Leaf) -> RpoDigest {
+        leaf.hash()
+    }
+}
+
+impl Default for NewSmt {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+// KEY
+// ================================================================================================
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct NewSmtKey {
+    word: Word,
+}
```

Comment:
Why not make this:
```
pub struct NewSmtKey(RpoDigest);
```
Then we don't need to implement `Ord` manually here.

Reply:
Done